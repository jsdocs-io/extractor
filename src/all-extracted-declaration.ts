import type {
  ExtractedClass,
  ExtractedClassConstructor,
  ExtractedClassMethod,
  ExtractedClassProperty,
} from "./extract-class";
import type { ExtractedEnum, ExtractedEnumMember } from "./extract-enum";
import type { ExtractedFunction } from "./extract-function";
import type {
  ExtractedInterface,
  ExtractedInterfaceCallSignature,
  ExtractedInterfaceConstructSignature,
  ExtractedInterfaceGetAccessor,
  ExtractedInterfaceIndexSignature,
  ExtractedInterfaceMethod,
  ExtractedInterfaceProperty,
  ExtractedInterfaceSetAccessor,
} from "./extract-interface";
import type { ExtractedNamespace } from "./extract-namespace";
import type { ExtractedTypeAlias } from "./extract-type-alias";
import type { ExtractedVariable } from "./extract-variable";

/**
`AllExtractedDeclaration` is the union of all possible declarations
that can be extracted, with some being found only in other declarations
(e.g., class method declarations are found only in a class declaration).
*/
export type AllExtractedDeclaration =
  | ExtractedVariable
  | ExtractedFunction
  | ExtractedClass
  | ExtractedClassConstructor
  | ExtractedClassProperty
  | ExtractedClassMethod
  | ExtractedInterface
  | ExtractedInterfaceProperty
  | ExtractedInterfaceMethod
  | ExtractedInterfaceConstructSignature
  | ExtractedInterfaceCallSignature
  | ExtractedInterfaceIndexSignature
  | ExtractedInterfaceGetAccessor
  | ExtractedInterfaceSetAccessor
  | ExtractedEnum
  | ExtractedEnumMember
  | ExtractedTypeAlias
  | ExtractedNamespace;

/**
`AllExtractedDeclarationKind` is the union of all discriminators
used to detect the kind of declaration.
*/
export type AllExtractedDeclarationKind = AllExtractedDeclaration["kind"];
