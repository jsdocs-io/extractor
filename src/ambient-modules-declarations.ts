import { ModuleDeclaration, Node, type Project } from "ts-morph";
import { isHidden } from "./is-hidden";
import { sourceFilePath } from "./source-file-path";

export type AmbientModulesDeclarationsReturn = {
  containerName: string;
  exportName: string;
  declaration: ModuleDeclaration;
}[];

export const ambientModulesDeclarations = (
  containerName: string,
  project: Project,
  pkgName?: string,
): AmbientModulesDeclarationsReturn => {
  const ambientModulesDeclarations = [];
  for (const symbol of project.getAmbientModules()) {
    for (const declaration of symbol.getDeclarations()) {
      if (isHidden(declaration) || !Node.isModuleDeclaration(declaration)) {
        continue;
      }
      if (pkgName && !sourceFilePath(declaration).startsWith(`/${pkgName}`)) {
        // Ignore ambient modules that are not from the analyzed package.
        continue;
      }
      const exportName = declaration.getName();
      // TODO:
      // Remove surrounding quotes and eventual spaces
      // const exportID = exportName
      //   .replace(/"|'/g, "")
      //   .replace(/\s/g, "_")
      //   .trim();
      ambientModulesDeclarations.push({
        containerName,
        exportName,
        declaration,
      });
    }
  }
  return ambientModulesDeclarations;
};
