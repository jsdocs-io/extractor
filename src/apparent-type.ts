import { TypeFormatFlags, ts, type Node } from "ts-morph";

export function apparentType(node: Node): string {
	// See:
	// https://github.com/dsherret/ts-morph/issues/453#issuecomment-427405736
	// https://github.com/dsherret/ts-morph/issues/453#issuecomment-667578386
	return node
		.getType()
		.getApparentType()
		.getText(
			node,
			ts.TypeFormatFlags.NoTruncation | TypeFormatFlags.UseAliasDefinedOutsideCurrentScope,
		)
		.replace(/^Number$/, "number")
		.replace(/^Boolean$/, "boolean")
		.replace(/^String$/, "string");
}
