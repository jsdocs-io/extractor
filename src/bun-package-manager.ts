import { Effect } from "effect";
import { execa } from "execa";
import { InstallPackageError, PackageManager } from "./package-manager";

/** @internal */
export const bunPackageManager = (bunPath = "bun") =>
	PackageManager.of({
		installPackage: ({ pkg, cwd }) =>
			Effect.gen(function* (_) {
				// Run `bun add <pkg> --verbose`.
				// See https://bun.sh/docs/cli/add.
				const { stdout } = yield* _(
					Effect.tryPromise({
						try: () => execa(bunPath, ["add", pkg, "--verbose"], { cwd }),
						catch: (e) => new InstallPackageError({ cause: e }),
					}),
				);

				// With verbose output on, bun prints one line per installed package
				// (e.g., "foo@1.0.0"), including all installed dependencies.
				// These lines are between the two delimiting lines found here:
				// https://github.com/oven-sh/bun/blob/972a7b7080bd3066b54dcb43e9c91c5dfa26a69c/src/install/lockfile.zig#L5369-L5370.
				const lines = stdout.split("\n");
				const beginHash = lines.findIndex((line) => line.startsWith("-- BEGIN SHA512/256"));
				const endHash = lines.findIndex((line) => line.startsWith("-- END HASH"));
				const installedPackages = lines.slice(beginHash + 1, endHash);
				return installedPackages;
			}),
	});
