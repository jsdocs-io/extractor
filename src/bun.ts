import { execa } from "execa";

/** `Bun` is a wrapper for the Bun package manager. */
export class Bun {
	#bunCmd: string;

	constructor(bunCmd = "bun") {
		this.#bunCmd = bunCmd.trim();
	}

	/**
	`add` installs the given package in the given directory and
	returns a list of all installed packages (e.g., `["foo@1.0.0"]`).

	@see {@link https://bun.sh/docs/cli/add}
	*/
	async add(pkg: string, cwd: string): Promise<string[]> {
		// Run `bun add <pkg> --verbose` in the `cwd`.
		const { stdout } = await execa(this.#bunCmd, ["add", pkg, "--verbose"], { cwd });

		// With verbose output on, bun prints one line per installed package
		// (e.g., "foo@1.0.0"), including all installed dependencies.
		// These lines are between the two delimiting lines found here:
		// https://github.com/oven-sh/bun/blob/972a7b7080bd3066b54dcb43e9c91c5dfa26a69c/src/install/lockfile.zig#L5369-L5370.
		const lines = stdout.split("\n");
		const beginHash = lines.findIndex((line) => line.startsWith("-- BEGIN SHA512/256"));
		const endHash = lines.findIndex((line) => line.startsWith("-- END HASH"));
		const installedPackages = lines.slice(beginHash + 1, endHash);
		return installedPackages;
	}
}
