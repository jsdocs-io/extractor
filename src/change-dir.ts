import { Result } from "neverthrow";
import { OsError } from "./errors";

export const changeDir = Result.fromThrowable(
  process.chdir,
  (e) => new OsError("failed to change directory", { cause: e }),
);
