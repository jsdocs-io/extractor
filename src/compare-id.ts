export type WithId = {
  id: string;
};

export const compareId = (a: WithId, b: WithId): number =>
  a.id.localeCompare(b.id);
