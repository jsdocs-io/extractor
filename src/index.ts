export {
  extractDeclarations,
  type ExtractDeclarationsOptions,
  type ExtractedDeclaration,
  type ExtractedDeclarationKind,
} from "./extract-declarations";
