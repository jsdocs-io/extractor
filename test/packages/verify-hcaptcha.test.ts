import { expect, test } from "vitest";
import { extractPackageApi } from "../../src";

test("verify-hcaptcha@1.0.0", async () => {
  expect(
    (
      await extractPackageApi({
        pkg: "verify-hcaptcha@1.0.0",
      })
    )._unsafeUnwrap(),
  ).toMatchSnapshot();
});
