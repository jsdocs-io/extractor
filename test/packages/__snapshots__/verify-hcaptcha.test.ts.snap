// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`verify-hcaptcha@1.0.0 1`] = `
{
  "declarations": [
    {
      "docs": [
        "/**
 * \`rawVerifyHcaptchaToken\` verifies with the hCaptcha API that the response token
 * obtained from a captcha challenge is valid and returns the raw hCaptcha response.
 *
 * @param token - required: the token obtained from a user with a captcha challenge
 * @param secretKey - required: the secret key for your account
 * @param siteKey - optional but recommended: the site key for the website hosting the captcha challenge
 * @param remoteIp - optional: the IP address of the user submitting the challenge
 *
 * @returns a {@link RawHcaptchaResponse} with the verification result
 */",
      ],
      "file": "/verify-hcaptcha/dist/index.d.ts",
      "id": "function.rawVerifyHcaptchaToken",
      "kind": "function",
      "line": 161,
      "name": "rawVerifyHcaptchaToken",
      "signature": "rawVerifyHcaptchaToken: ({
  token,
  secretKey,
  siteKey,
  remoteIp,
}: {
  token: string;
  secretKey: string;
  siteKey?: string;
  remoteIp?: string;
}) => Promise<RawHcaptchaResponse>;",
    },
    {
      "docs": [
        "/**
 * \`verifyHcaptchaToken\` verifies with the hCaptcha API that the response token
 * obtained from a captcha challenge is valid.
 *
 * @param token - required: the token obtained from a user with a captcha challenge
 * @param secretKey - required: the secret key for your account
 * @param siteKey - optional but recommended: the site key for the website hosting the captcha challenge
 * @param remoteIp - optional: the IP address of the user submitting the challenge
 *
 * @returns a {@link HcaptchaResponse} with the verification result
 */",
      ],
      "file": "/verify-hcaptcha/dist/index.d.ts",
      "id": "function.verifyHcaptchaToken",
      "kind": "function",
      "line": 144,
      "name": "verifyHcaptchaToken",
      "signature": "verifyHcaptchaToken: ({
  token,
  secretKey,
  siteKey,
  remoteIp,
}: {
  token: string;
  secretKey: string;
  siteKey?: string;
  remoteIp?: string;
}) => Promise<HcaptchaResponse>;",
    },
    {
      "callSignatures": [],
      "constructSignatures": [],
      "docs": [
        "/**
 * \`HcaptchaResponse\` represents the response to the verification challenge
 * performed by calling {@link verifyHcaptchaToken}.
 *
 * @see {@link https://docs.hcaptcha.com/#verify-the-user-response-server-side}
 */",
      ],
      "file": "/verify-hcaptcha/dist/index.d.ts",
      "getAccessors": [],
      "id": "interface.HcaptchaResponse",
      "indexSignatures": [],
      "kind": "interface",
      "line": 62,
      "methods": [],
      "name": "HcaptchaResponse",
      "properties": [
        {
          "docs": [
            "/**
     * Optional: UTC timestamp of the challenge in ISO 8601 format
     * (for example, \`2021-10-02T18:12:10.149Z\`)
     */",
          ],
          "file": "/verify-hcaptcha/dist/index.d.ts",
          "id": "interface.HcaptchaResponse.property.challengeTimestamp",
          "kind": "interface-property",
          "line": 72,
          "name": "challengeTimestamp",
          "signature": "readonly challengeTimestamp?: string;",
        },
        {
          "docs": [
            "/** Optional: true if the response will be credited */",
          ],
          "file": "/verify-hcaptcha/dist/index.d.ts",
          "id": "interface.HcaptchaResponse.property.credit",
          "kind": "interface-property",
          "line": 76,
          "name": "credit",
          "signature": "readonly credit?: boolean;",
        },
        {
          "docs": [
            "/**
     * Optional: list of error codes
     *
     * @see {@link HcaptchaError}
     */",
          ],
          "file": "/verify-hcaptcha/dist/index.d.ts",
          "id": "interface.HcaptchaResponse.property.errorCodes",
          "kind": "interface-property",
          "line": 82,
          "name": "errorCodes",
          "signature": "readonly errorCodes?: HcaptchaError[];",
        },
        {
          "docs": [
            "/** Optional: hostname of the website where the challenge was solved */",
          ],
          "file": "/verify-hcaptcha/dist/index.d.ts",
          "id": "interface.HcaptchaResponse.property.hostname",
          "kind": "interface-property",
          "line": 74,
          "name": "hostname",
          "signature": "readonly hostname?: string;",
        },
        {
          "docs": [
            "/** Enterprise-only feature: score for malicious activity */",
          ],
          "file": "/verify-hcaptcha/dist/index.d.ts",
          "id": "interface.HcaptchaResponse.property.score",
          "kind": "interface-property",
          "line": 84,
          "name": "score",
          "signature": "readonly score?: number;",
        },
        {
          "docs": [
            "/** Enterprise-only feature: list of reasons for the malicious activity score */",
          ],
          "file": "/verify-hcaptcha/dist/index.d.ts",
          "id": "interface.HcaptchaResponse.property.scoreReasons",
          "kind": "interface-property",
          "line": 86,
          "name": "scoreReasons",
          "signature": "readonly scoreReasons?: string[];",
        },
        {
          "docs": [
            "/**
     * True if the token is valid and meets the specified security criteria
     * (for example, if the site key is associated to the secret key)
     */",
          ],
          "file": "/verify-hcaptcha/dist/index.d.ts",
          "id": "interface.HcaptchaResponse.property.success",
          "kind": "interface-property",
          "line": 67,
          "name": "success",
          "signature": "readonly success: boolean;",
        },
      ],
      "setAccessors": [],
      "signature": "interface HcaptchaResponse {}",
    },
    {
      "callSignatures": [],
      "constructSignatures": [],
      "docs": [
        "/**
 * \`RawHcaptchaResponse\` represents the raw response to the verification challenge
 * obtained by directly calling the hCaptcha API endpoint
 * with {@link rawVerifyHcaptchaToken}.
 *
 * @see {@link https://docs.hcaptcha.com/#verify-the-user-response-server-side}
 */",
      ],
      "file": "/verify-hcaptcha/dist/index.d.ts",
      "getAccessors": [],
      "id": "interface.RawHcaptchaResponse",
      "indexSignatures": [],
      "kind": "interface",
      "line": 124,
      "methods": [],
      "name": "RawHcaptchaResponse",
      "properties": [
        {
          "docs": [],
          "file": "/verify-hcaptcha/dist/index.d.ts",
          "id": "interface.RawHcaptchaResponse.property."error-codes"",
          "kind": "interface-property",
          "line": 129,
          "name": ""error-codes"",
          "signature": "readonly "error-codes"?: string[];",
        },
        {
          "docs": [],
          "file": "/verify-hcaptcha/dist/index.d.ts",
          "id": "interface.RawHcaptchaResponse.property.challenge_ts",
          "kind": "interface-property",
          "line": 126,
          "name": "challenge_ts",
          "signature": "readonly challenge_ts?: string;",
        },
        {
          "docs": [],
          "file": "/verify-hcaptcha/dist/index.d.ts",
          "id": "interface.RawHcaptchaResponse.property.credit",
          "kind": "interface-property",
          "line": 128,
          "name": "credit",
          "signature": "readonly credit?: boolean;",
        },
        {
          "docs": [],
          "file": "/verify-hcaptcha/dist/index.d.ts",
          "id": "interface.RawHcaptchaResponse.property.hostname",
          "kind": "interface-property",
          "line": 127,
          "name": "hostname",
          "signature": "readonly hostname?: string;",
        },
        {
          "docs": [],
          "file": "/verify-hcaptcha/dist/index.d.ts",
          "id": "interface.RawHcaptchaResponse.property.score",
          "kind": "interface-property",
          "line": 130,
          "name": "score",
          "signature": "readonly score?: number;",
        },
        {
          "docs": [],
          "file": "/verify-hcaptcha/dist/index.d.ts",
          "id": "interface.RawHcaptchaResponse.property.score_reason",
          "kind": "interface-property",
          "line": 131,
          "name": "score_reason",
          "signature": "readonly score_reason?: string[];",
        },
        {
          "docs": [],
          "file": "/verify-hcaptcha/dist/index.d.ts",
          "id": "interface.RawHcaptchaResponse.property.success",
          "kind": "interface-property",
          "line": 125,
          "name": "success",
          "signature": "readonly success: boolean;",
        },
      ],
      "setAccessors": [],
      "signature": "interface RawHcaptchaResponse {}",
    },
    {
      "docs": [
        "/**
 * \`HcaptchaError\` collects the errors explaining why a verification challenge failed.
 *
 * @see {@link HcaptchaResponse}
 * @see {@link https://docs.hcaptcha.com/#siteverify-error-codes-table}
 */",
      ],
      "file": "/verify-hcaptcha/dist/index.d.ts",
      "id": "type.HcaptchaError",
      "kind": "type",
      "line": 94,
      "name": "HcaptchaError",
      "signature": "type HcaptchaError =
  /** Secret key is missing */
  | "missing-input-secret"
  /** Secret key is invalid */
  | "invalid-input-secret"
  /** User response token is missing */
  | "missing-input-response"
  /** User response token is invalid */
  | "invalid-input-response"
  /** Site key is invalid */
  | "invalid-sitekey"
  /** Remote user IP is invalid */
  | "invalid-remoteip"
  /** Request is invalid */
  | "bad-request"
  /** User response token is invalid or has already been checked */
  | "invalid-or-already-seen-response"
  /** Must use the test site key when using a test verification token */
  | "not-using-dummy-passcode"
  /** Must use the test secret key when using a test verification token */
  | "not-using-dummy-secret"
  /** The site key is not associated to the secret key */
  | "sitekey-secret-mismatch";",
    },
  ],
  "name": "verify-hcaptcha",
  "overview": "/**
 *
 * A no dependencies, fully typed library to verify hCaptcha tokens
 * submitted by users when solving CAPTCHA challenges.
 *
 * @remarks
 *
 * Note: this is an unofficial library; we are not affiliated with hCaptcha.com
 *
 * @example
 *
 * Verify a token submitted by a user:
 *
 * \`\`\`typescript
 * import { verifyHcaptchaToken } from 'verify-hcaptcha';
 *
 * (async () => {
 *     const result = await verifyHcaptchaToken({
 *       token: "USER-SUBMITTED-RESPONSE-TOKEN",
 *       secretKey: "YOUR-SECRET-KEY",
 *       siteKey: "YOUR-SITE-KEY",
 *     });
 *
 *     if (result.success) {
 *       console.log("User is human");
 *     } else {
 *       console.log("User is robot");
 *     }
 * })();
 * \`\`\`
 *
 * @example
 *
 * Verify a token submitted by a user and get the raw response from hCaptcha:
 *
 * \`\`\`typescript
 * import { rawVerifyHcaptchaToken } from 'verify-hcaptcha';
 *
 * (async () => {
 *     const result = await rawVerifyHcaptchaToken({
 *       token: "USER-SUBMITTED-RESPONSE-TOKEN",
 *       secretKey: "YOUR-SECRET-KEY",
 *       siteKey: "YOUR-SITE-KEY",
 *     });
 *
 *     if (result.success) {
 *       console.log("User is human");
 *     } else {
 *       console.log("User is robot");
 *     }
 * })();
 * \`\`\`
 *
 * @packageDocumentation
 */",
  "packages": [
    "verify-hcaptcha@1.0.0",
  ],
  "subpath": ".",
  "types": "dist/index.d.ts",
  "version": "1.0.0",
}
`;
